prov/verbs/src/verbs_ep.c

```diff
@@ -1544,13 +1544,14 @@ static int vrb_srq_close(fid_t fid)
 {
 	struct vrb_srq_ep *srq_ep = container_of(fid, struct vrb_srq_ep,
 						 ep_fid.fid);
+	struct vrb_cq *cq = srq_ep->xrc.cq;
 	int ret;
 
 	if (srq_ep->domain->flags & VRB_USE_XRC) {
-		if (srq_ep->xrc.cq) {
-			fastlock_acquire(&srq_ep->xrc.cq->xrc.srq_list_lock);
+		if (cq) {
+			fastlock_acquire(&cq->xrc.srq_list_lock);
 			ret = vrb_xrc_close_srq(srq_ep);
-			fastlock_release(&srq_ep->xrc.cq->xrc.srq_list_lock);
+			fastlock_release(&cq->xrc.srq_list_lock);
```

Comment:
I'm not sure the commit message is correct.  The problem that I see is that vrb_xrc_close_srq() does:
```
srq_ep->xrc.cq = NULL;
```
which results in a NULL dereference calling fastlock_release.  Must first question is why does vrb_xrc_close_srq() set that to NULL to begin with?  That looks like a bug.

Reply:
That is the problem, at least what I'm avoiding. The XRC SRQ and CQ can be closed in any order. Unlike a traditional SRQ, the XRC SRQ references the RX CQ directly (not via the QP RX CQ). This pointer just acts as flag, indicating that the CQ has been closed and it closed the associated XRC SRQ, so the hardware XRC SRQ doesn't need to be done when the shared RX context is closed.  In the case of the error, the SRQ is closed first, it removes itself from the CQ, but hits the segfault because that is done while holding the CQ XRC SRQ list lock. As RxM is the current consumer of XRC, and it closes the CQ first, this error is not seen.

Reply:
We should remove setting cq = NULL from xrc_close_srq().  That's an unintended side effect that doesn't make sense.

The CQ maintains the SRQ list and the lock.  It seems that there is a required ordering here for the SRQ to be closed and cleaned up from the CQ before the CQ is itself closed.  Allowing an arbitrary order looks like a recipe for errors.

Reply:
@shefty - I agree that is the preferred/natural ordering. I originally allowed that because RxM closed the CQ first which works with standard SRQs, and there is nothing that dictates in the Libfabric API that the shared RX context should be closed prior to the CQ. With that said though, if you are OK with it, it simplify the code if I switch RxM to close the require the SRQ be closed first and will make it more maintainable. If you are good with that, I will do that and update the PR.